scripts/download_artifacts.py

```diff
@@ -272,8 +285,98 @@ def issue_hashes(repo_name: str, token: str):
     return hashes
 
+def search_and_download_previous_report_artifact(
+    artifact_name_template:str, previous_hash: str, prev_commits: List[str], repo_name: str, token: str
+):
+    """Download a most recent previous report artifact and return the corresponding hash.
+       Return None if no corresponding hash has been found
+    """
+    artifact_name_template += "-report.log"
+
+    # check if we already have a previous artifact available
+    # mostly for regenerate issues
+    if previous_hash:
+        name_components = artifact_name_template.split("-")
+        name_components[4] = "{}"
+        previous_name = "-".join(name_components).format(previous_hash)
+        name_components[4] = "[^-]*"
+        name_regex = "-".join(name_components)
+        dir_contents = " ".join(os.listdir("./previous_logs"))
+        possible_previous_logs = re.findall(name_regex, dir_contents)
+        if len(possible_previous_logs) > 1:
+            print(
+                f"found more than 1 previous log for {name_regex}: {possible_previous_logs}"
+            )
+            for log in possible_previous_logs:  # remove non-recent logs
+                if previous_name not in log:
+                    print(f"removing {log} from previous_logs")
+                    os.remove(os.path.join("./previous_logs", log))
+            return previous_hash
+        if len(possible_previous_logs) == 1:
+            print(
+                f"found single log: {possible_previous_logs[0]}. Skipping download"
+            )
+            return previous_hash    
+
+    # download previous artifact
+    base_hash, base_id = get_valid_artifact_hash(
+        prev_commits, repo_name, token, artifact_name_template
+    )
+    if base_hash != "No valid hash":
+        artifact_zip = download_artifact(
+            artifact_name_template.format(base_hash), str(base_id), token, repo_name, "./temp/"
+        )
+        extract_artifact(
+            artifact_zip,
+            outdir="previous_logs",
+        )
+        return base_hash
+
+    print(
+        f"found no valid hash for {artifact_name_template}. possible hashes: {prev_commits}"
+    )
+    return None
+
+def download_build_log_artifact(
+    artifact_name_template:str, base_hash: str, repo_name: str, token: str, output_dir: str
+):
+    """Download the build log zipfiles for a corresponding base_hash and extract them to output_dir"""
+    # input validation
+    if not base_hash:
+        return
+
+    target_format_pat = re.compile(r"^gcc-")
+    BUILD_LOG_SUFFIX = "-build-log"
+    # Remove the starting gcc- and concatenate with the suffix
+    artifact_name = target_format_pat.sub('', artifact_name_template.format(base_hash) + BUILD_LOG_SUFFIX)
+
+    # Check if the artifact name exists inside the output directory
+    previous_build_log_path = Path(f'./{output_dir}/{artifact_name}')
+    if previous_build_log_path.exists():
```

Comment:
just wondering, will this condition ever be true? I'm assuming this is the only location we download the build logs. Not sure if it'll be downloaded though a github action or something

Reply:
It could be true in the future if more scripts are added for the build log. Since it doesn't require much computation for a check, it would be nice to have a check here.

Reply:
I'd be tempted to put it in as an assert if it's currently unexpected. That way we get notified about the changing behavior and can investigate.

Reply:
I believe the assertion is unnecessary because having an already downloaded file is not an error. The primary purpose of this check is to prevent duplicate downloads and excessive GitHub API calls. 

  Users who might download the build log multiple times should implement this to minimize API calls. Given that file existence checks are computationally inexpensive, maintaining this check is beneficial.

Reply:
It's not an error but I don't/wouldn't expect it to be true and I'd be interested in investigating cases where this does prevent a duplicate download.
If it ends up being tripped in an intended way I'd be ok with downgrading it to a if-gate as it's currently implemented here.

Reply:
I see if you want to check remaining duplicate for now then I can write an assertion with a comment letting the future user to remove it when their test fails